commit 5816339310b2d9623cf413d33e538b45e815da5d
Author: David S. Miller <davem@davemloft.net>
Date:   Wed May 7 02:24:28 2008 -0700

    sparc: Fix mmap VA span checking.
    
    We should not conditionalize VA range checks on MAP_FIXED.
    
    Signed-off-by: David S. Miller <davem@davemloft.net>

Adjusted to apply to Debian's 2.6.18 by dann frazier <dannf@hp.com>

diff -urpN linux-source-2.6.18.orig/arch/sparc/kernel/sys_sparc.c linux-source-2.6.18/arch/sparc/kernel/sys_sparc.c
--- linux-source-2.6.18.orig/arch/sparc/kernel/sys_sparc.c	2006-09-19 21:42:06.000000000 -0600
+++ linux-source-2.6.18/arch/sparc/kernel/sys_sparc.c	2008-05-23 10:13:30.000000000 -0600
@@ -223,8 +223,7 @@ int sparc_mmap_check(unsigned long addr,
 {
 	if (ARCH_SUN4C_SUN4 &&
 	    (len > 0x20000000 ||
-	     ((flags & MAP_FIXED) &&
-	      addr < 0xe0000000 && addr + len > 0x20000000)))
+	     (addr < 0xe0000000 && addr + len > 0x20000000)))
 		return -EINVAL;
 
 	/* See asm-sparc/uaccess.h */
diff -urpN linux-source-2.6.18.orig/arch/sparc64/kernel/sys_sparc.c linux-source-2.6.18/arch/sparc64/kernel/sys_sparc.c
--- linux-source-2.6.18.orig/arch/sparc64/kernel/sys_sparc.c	2006-09-19 21:42:06.000000000 -0600
+++ linux-source-2.6.18/arch/sparc64/kernel/sys_sparc.c	2008-05-23 10:13:30.000000000 -0600
@@ -555,13 +555,13 @@ int sparc64_mmap_check(unsigned long add
 		if (len >= STACK_TOP32)
 			return -EINVAL;
 
-		if ((flags & MAP_FIXED) && addr > STACK_TOP32 - len)
+		if (addr > STACK_TOP32 - len)
 			return -EINVAL;
 	} else {
 		if (len >= VA_EXCLUDE_START)
 			return -EINVAL;
 
-		if ((flags & MAP_FIXED) && invalid_64bit_range(addr, len))
+		if (invalid_64bit_range(addr, len))
 			return -EINVAL;
 	}
 
